crates/driver/src/tests/cases/verify_asset_flow/liquidity.rs

```diff
@@ -0,0 +1,478 @@
+            liquidity_orders: |sell_token: eth::H160, buy_token: eth::H160| {
+                vec![LiquidityOrder {
+                    sell: eth::Asset {
+                        amount: 50.into(),
+                        token: sell_token.into(),
+                    },
+                    buy: eth::Asset {
+                        amount: 50.into(),
+                        token: buy_token.into(),
+                    },
+                    side: competition::order::Side::Buy,
+                    executed: 50.into(),
+                    partially_fillable: false,
+                }]
+            },
+        },
+        TestCase {
+            // Single buy order which is partially fillable.
+            valid: true,
+            liquidity_orders: |sell_token: eth::H160, buy_token: eth::H160| {
+                vec![LiquidityOrder {
+                    sell: eth::Asset {
+                        amount: 100.into(),
+                        token: sell_token.into(),
+                    },
+                    buy: eth::Asset {
+                        amount: 100.into(),
+                        token: buy_token.into(),
+                    },
+                    side: competition::order::Side::Buy,
+                    executed: 50.into(),
+                    partially_fillable: true,
+                }]
+            },
+        },
+        TestCase {
+            // Partially fillable sell and buy order.
+            valid: true,
+            liquidity_orders: |sell_token: eth::H160, buy_token: eth::H160| {
+                vec![
+                    // This order ends up selling 40 and buying 30.
+                    LiquidityOrder {
+                        sell: eth::Asset {
+                            amount: 80.into(),
+                            token: sell_token.into(),
+                        },
+                        buy: eth::Asset {
+                            amount: 60.into(),
+                            token: buy_token.into(),
+                        },
+                        side: competition::order::Side::Buy,
+                        executed: 30.into(),
+                        partially_fillable: true,
+                    },
+                    // This order ends up selling 10 and buying 20.
+                    LiquidityOrder {
+                        sell: eth::Asset {
+                            amount: 100.into(),
+                            token: sell_token.into(),
+                        },
+                        buy: eth::Asset {
+                            amount: 200.into(),
+                            token: buy_token.into(),
+                        },
+                        side: competition::order::Side::Sell,
+                        executed: 10.into(),
+                        partially_fillable: true,
+                    },
+                ]
+            },
+        },
+        TestCase {
+            // Single order which attempts to buy too much.
```

Comment:
I'm not sure why this case should fail. It would sell `5` to get `100` in return which would be a great deal leaving us with `45 sell_token` and `50 buy_token` more than expected.

Reply:
Could it be that the asset flow test requires / assumes that tokens always have the same price? With that assumption it would make sense that this case fails as it would produce `buy_tokens` out of thin air when no equivalent amount of `sell_tokens` was offered in return. However if a solver actually finds market maker that is willing to sell tokens for way below market value this test should probably not fail. Not sure if I'm overlooking something or the underlying test needs special treatment for liquidity orders (and maybe even for limit orders as they are also not guaranteed to trade at the same price).

Reply:
I think the issue here is that you would buy too much. Specifically, only 50 of the buy tokens are "avaiable" for the liquidity order (going in the same direction as the user order), but we are transferring 100 of buy token out for this liquidity order, so I would expect a flux of -50 buy_token (hence not respecting token conservation).

Reply:
Ahh, right. I got the "flux" backwards. 🤦‍♂️